CP-32622: Remove 'select' from stdext

edwintorok opened the pull request:
Draft, because I need to add some unit tests here.

I'd also like to add a separate function that works only on sockets which is much simpler and uses just 'setsockopt_float'.

snwoods committed:
Add unit tests for threadext

Signed-off-by: Steven Woods <steven.woods@citrix.com>
Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
Unix.time_limited_write: fix timeout behaviour on >64KiB writes/reads

`Unix.write` would internally loop until the desired number of bytes are sent,
or `EAGAIN`/`EWOULDBLOCK`/another error is reached.
It cannot check for timeouts because it is not aware that we'd want one.

For pipes and sockets in non-blocking mode this wouldn't be a problem, but this function
is often called with block devices too.
However according to `pselect(3p)` it is a no-op on regular files:
"File descriptors associated with regular files shall always select true for ready to read, ready to write, and error conditions."
And the behaviour on block devices is left unspecified by POSIX, and `select(2)` on Linux doesn't document the behaviour either:
"The pselect() and select() functions shall support regular files, terminal and pseudo‐terminal devices, STREAMS‐based files, FIFOs, pipes, and sockets. The behavior of pselect() and select() on file descriptors that refer to other types of file is unspecified"

Although timeouts for a completely stuck block device cannot be implemented, we can still implement timeouts for a *slow* block device.
Use `Unix.single_{write,read}` instead which gives full control of the iteration to the caller.

The only way to interrupt stuck IO on a block device or regular file would be to use a separate process and `SIGKILL` it after a timeout (this is what `block_device_io` in XAPI does).

These approaches do not work:
* `alarm` or `setitimer` would only interrupt one thread in a multi-threaded program.
* `pthread_kill` can be used to send a signal to a particular thread, but that requires `EINTR` behaviour on syscalls to be enabled
* XAPI expects `SA_RESTART` semantics from syscalls, and would fail an assertion if it ever sees `EINTR` in some paths,
so although the syscall *would* get interrupted, it'd also immediately resume without giving the caller a chance to check for timeouts
* even if it'd work there are no bindings to `pthread_kill` in OCaml

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
Unix.time_limited_{read,write}: replace select with Polly

'select' has a hardcoded limit of 1024 file descriptors.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-32622: replace select with Thread.delay

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-32622: Delay: replace select with time_limited_read

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-32622: replace select in proxy with polly

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

lindig commented on lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -409,26 +409,52 @@ let string_of_signal x =
+  with_polly @@ fun polly ->
 
   try
     while true do
```
could this be a recursive function with `any` as a parameter which is used for termination?

lindig commented on lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -501,49 +527,68 @@ let really_write_string fd string =
 
+let to_milliseconds ms = ms *. 1000. |> int_of_float
+
+(* Allocating a new polly and waiting like this results in at least 3 system calls.
```
Needs reformat to break long lines

edwintorok replied:
reformat PR here: https://github.com/xapi-project/stdext/pull/82
I then installed `merge-fmt` and rebased this PR as documented here: https://github.com/hhugo/merge-fmt#getting-started
I didn't have to manually solve any merge conflicts and the PR was almost formatted correctly too (when run reformat at the end, there was only one small fixup to the unit test commit, which I squashed into it).

Although nothing has actually changed wrt to this line since it is a comment (and comments aren't reformatted here), so I manually fixed up the comment

lindig commented on xapi-stdext-unix-opam.template:
```diff
@@ -0,0 +1,2 @@
+depexts: ["linux-headers"] {os-distribution = "alpine"}
+available: [ os = "macos" | os = "linux" ]
```
epoll is not available on macOS; not sure if this is relevant here.

edwintorok replied:
We could use `iomux` if needed, it supports more OSes. If we want to make `xe` available on more platforms then we should either remove its dependency on stdext, or avoid hard dependencies on epoll.
For now I've updated the template to drop macos.

edwintorok replied:
Although contrary to its claim in the package description 'epoll' support isn't actually implement in `iomux` yet, it is just `poll`.

edwintorok committed:
add Unixext.time_limited_single_read

It is too easy to misuse Unixext.time_limited_read because that one takesan absolute timestamp
as parameter, not a relative one.

Introduce a new function that takes a relative time as parameter, and doesn't loop.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
I hadn't noticed that there is already https://github.com/xapi-project/stdext/pull/71 which also has one unit test.
I cherry picked the unit test to this PR and it found a bug: `time_limited_read` takes an absolute time not relative time as timeout (which means that it always timed out instantly because it interpreted the relative timeout as an absolute time in the distant past).

That is very error prone, so I introduced a `time_limited_single_read` which takes a relative time as parameter, and this time it is labeled so we know what it is.

edwintorok committed:
[maintenance]: run ocamlformat

'make format' with ocamlformat 0.22.4

No functional change.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

psafont commented:
> I cherry picked the unit test to this PR and it found a bug: time_limited_read takes an absolute time not relative time as timeout (which means that it always timed out instantly because it interpreted the relative timeout as an absolute time in the distant past).

> That is very error prone, so I introduced a time_limited_single_read which takes a relative time as parameter, and this time it is labeled so we know what it is.

I'll try to find out some time to add a timer module I've been developing lately to module in the date / time package that can be used for timed functions:

```ocaml
module Timer : sig
  (** This module is useful for knowing that a set amount of time has passed
    after some event happened. For example, to know when pasta is cooked al
    dente. *)
  type t

  type remaining = Spare of Mtime.Span.t | Excess of Mtime.Span.t

  val start : timeout:Mtime.Span.t -> t
  (** [start ~timeout] starts a timer that expires after [timeout] has passed.
      The wait is done in monotonic time, not in POSIX time. *)

  val timeout : t -> Mtime.Span.t
  (** [timeout timer] returns the amount of time after which the timer expires,
      from the moment it was started. *)

  val expired : t -> bool
  (** [expired timer] returns whether [timer] has reached the timeout it was
      set to. *)

  val elapsed : t -> Mtime.Span.t
  (** [elapsed timer] returns the amount of time elapsed since [timer] was
      started. *)

  val remaining : t -> remaining
  (** [remaining timer] returns the amount of spare time is left until it
      expires, or the amount of excess time since it expired. *)

  val deadline_of : t -> float
  (** [deadline_of timer] returns the posix timestamp when the timer expires.
      This is an approximation as the timer doesn't take leap seconds into
      account when waiting. The use of this function is discouraged and it's
      only provided for backwards-compatible reasons. *)
end
```

edwintorok commented:
We'd also need to convert that into whatever form the underlying POSIX API expects, and they may use different clocks.
E.g. 'epoll_wait' (and thus polly) will use monotonic clocks, so that'll match what mtime uses, but XAPI computes a deadline using 'gettimeofday'.

Shouldn't we just merge stdext into XAPI itself? it looks like we might want to refactor some of the APIs, and having them in separate repos like this makes that more difficult (I'd prefer if XAPI used mtime too, but we cannot do that easily without breaking changes, whereas in a monorepo it'd be a lot simpler).

edwintorok commented:
Continued in https://github.com/xapi-project/xen-api/pull/5402
